p2panda-rs/src/entry/entry_signed.rs

```diff
@@ -176,10 +192,160 @@ mod tests {
         assert_eq!(entry_signed_encoded.payload_hash(), expected_payload_hash)
     }
 
-    #[test]
-    fn validate() {
-        // Invalid hex string
-        assert!(EntrySigned::new("123456789Z").is_err());
+    #[rstest]
+    #[case::valid_first_entry(entry_signed_encoded_unvalidated(
+        1,
+        1,
+        None,
+        None,
+        Some(operation(Some(operation_fields(default_fields())), None, None)),
+        key_pair(DEFAULT_PRIVATE_KEY)
+    ))]
+    #[case::valid_entry_with_backlink(entry_signed_encoded_unvalidated(
+        2,
+        1,
+        Some(random_hash()),
+        None,
+        Some(operation(Some(operation_fields(default_fields())), None, None)),
+        key_pair(DEFAULT_PRIVATE_KEY)
+    ))]
+    #[case::valid_entry_with_skiplink_and_backlink(entry_signed_encoded_unvalidated(
+        13,
+        1,
+        Some(random_hash()),
+        Some(random_hash()),
+        Some(operation(Some(operation_fields(default_fields())), None, None)),
+        key_pair(DEFAULT_PRIVATE_KEY)
+    ))]
+    #[case::skiplink_ommitted_when_sam_as_backlink(entry_signed_encoded_unvalidated(
+        14,
+        1,
+        Some(random_hash()),
+        None,
+        Some(operation(Some(operation_fields(default_fields())), None, None)),
+        key_pair(DEFAULT_PRIVATE_KEY)
+    ))]
+
+    fn validate(#[case] entry_signed_encoded_unvalidated: String) {
+        assert!(EntrySigned::new(&entry_signed_encoded_unvalidated).is_ok());
+    }
+
+    #[rstest]
+    #[case::empty_string("", "Bytes to decode had length of 0")]
+    #[case::invalid_hex_string("123456789Z", "invalid hex encoding in entry")]
+    #[case::another_invalid_hex_string(":{][[5£$%*(&*££  ++`/.", "invalid hex encoding in entry")]
+    #[case::seq_number_zero(
+        entry_signed_encoded_unvalidated(
+            0,
+            1,
+            None,
+            None,
+            Some(operation(Some(operation_fields(default_fields())), None, None)),
+            key_pair(DEFAULT_PRIVATE_KEY)
+        ),
+        "Entry sequence must be larger than 0 but was 0"
+    )]
+    // TODO: This doesn't error...
+    //
+    // #[case::log_id_zero(
+    //     entry_signed_encoded_unvalidated(
+    //         1,
+    //         0,
+    //         None,
+    //         None,
+    //         Some(operation(Some(operation_fields(default_fields())), None, None)),
+    //         key_pair(DEFAULT_PRIVATE_KEY)
+    //     ),
+    //     "Log id must be larger than 0 but was 0"
+    // )]
```

Comment:
:point_up:

Reply:
https://github.com/p2panda/p2panda/issues/369

Reply:
This is valid, we want log_id = 0, but we need to update the spec to reflect this.